cartographer_ros/src/cartographer_node_main.cc

```diff
@@ -90,6 +93,18 @@ Rigid3d ToRigid3d(const geometry_msgs::TransformStamped& transform) {
                                     transform.transform.rotation.z));
 }
 
+Rigid3d ToRigid3d(const geometry_msgs::Pose& pose) {
+  return Rigid3d(
+      Eigen::Vector3d(pose.position.x, pose.position.y, pose.position.z),
+      Eigen::Quaterniond(pose.orientation.w, pose.orientation.x,
+                         pose.orientation.y, pose.orientation.z));
+}
+
+PoseCovariance ToPoseCovariance(boost::array<double, 36ul> covariance) {
```

Comment:
1. Seems similar in behavior to std::array, so 'covariance' is passed by value. Instead of calling data() in this function, pass as "const double* covariance"? Alternatively, "const boost::array<double, 36>& covariance".
2. Return value is PoseCovariance, so the Eigen::Map<> is used to construct an Eigen::Matrix, i.e. it is a copy of the data. No problem with lifetimes.